### PR TITLE
osd: fix reference leak when ManifestOp is not used

### PR DESCRIPTION
**Pull request by myoungwon**
fixes: https://tracker.ceph.com/issues/50299

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -3580,6 +3580,8 @@ bool PrimaryLogPG::inc_refcount_by_set(OpContext* ctx, object_manifest_t& set_ch
     if (mop->tids.size()) {
       manifest_ops[ctx->obs->oi.soid] = mop;
       manifest_ops[ctx->obs->oi.soid]->op = ctx->op;
+    } else {
+      delete mop->cb;
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc:3584]: why not call `delete cb`  right in the ctor of `ManifestOp`?

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -3578,9 +3578,10 @@ bool PrimaryLogPG::inc_refcount_by_set(OpContext* ctx, object_manifest_t& set_ch
       }
     }
     if (mop->tids.size()) {
+      mop->cb = new RefCountCallback(ctx, osd_op);
       manifest_ops[ctx->obs->oi.soid] = mop;
       manifest_ops[ctx->obs->oi.soid]->op = ctx->op;
-    }
+    }
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc:3584]: might want to remove the trailing space.

Comment (myoungwon) [reply src/osd/PrimaryLogPG.cc]: Thanks. Fixed.

File: src/osd/PrimaryLogPG.cc [hunk 3]
```diff
@@ -3539,7 +3539,7 @@ bool PrimaryLogPG::inc_refcount_by_set(OpContext* ctx, object_manifest_t& set_ch
     refs);
   bool need_inc_ref = false;
   if (!refs.is_empty()) {
-    ManifestOpRef mop = std::make_shared<ManifestOp>(new RefCountCallback(ctx, osd_op));
+    ManifestOpRef mop(std::make_shared<ManifestOp>(nullptr));
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:3542]: Give ManifestOp a default constructor and use it here.  Remove this one if there are no other users.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/osd/PrimaryLogPG.h [hunk 4]
```diff
@@ -260,7 +260,7 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
   friend struct RefCountCallback;
   struct ManifestOp {
     RefCountCallback *cb;
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.h:262]: cb = nullptr

File: src/osd/PrimaryLogPG.h [hunk 5]
```diff
@@ -270,7 +270,9 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
     
 
     ManifestOp(RefCountCallback* cb)
-      : cb(cb), objecter_tid(0) {}
+      : cb(cb) {}
+    ManifestOp()
+      : cb(nullptr) {}
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.h:275]: ManifestOp() = default;

See above.

**Review by athanatos — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-04-17_15:15:57-rados-wip-kefu-testing-2021-04-16-2148-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-19_07:13:12-rados-wip-kefu-testing-2021-04-19-1211-distro-basic-smithi/

**Pull request merged**